coap/adapter.go

```diff
@@ -1,61 +1,145 @@
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
-}
+// Recieve method processes message and pushes it to NATS.
+func (ca *AdapterService) Recieve(auth AuthProvider) MsgHandler {
+	return func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message {
+		ca.logger.Log("message", fmt.Sprintf("Got message in Recieve: path=%q: %#v from %v", m.Path(), m, a))
+		var res *gocoap.Message
+
+		if m.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: m.MessageID,
+				Token:     m.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
 
-// BridgeHandler functions is a handler for messages recieved via NATS
-func (ca *CoAPAdapter) BridgeHandler(nm *broker.Msg) {
-	log.Printf("Received a message: %s\n", string(nm.Data))
+		if len(m.Payload) == 0 && m.IsConfirmable() {
+			res.Code = gocoap.BadRequest
+			return res
+		}
+
+		cid := mux.Var(m, channel)
+		publisher, err := authorize(m, res, cid, auth)
+		if err != nil {
+			ca.logger.Log("error", fmt.Sprintf("%s", err))
+			return res
+		}
 
-	// And write it into the database
-	m := mainflux.RawMessage{}
-	if len(nm.Data) > 0 {
-		if err := json.Unmarshal(nm.Data, &m); err != nil {
-			log.Println("Can not decode adapter msg")
-			return
+		n := mainflux.RawMessage{
+			Channel:   cid,
+			Publisher: publisher,
+			Protocol:  protocol,
+			Payload:   m.Payload,
 		}
+
+		if err := ca.pubSub.Publish(n); err != nil {
+			if m.IsConfirmable() {
+				res.Code = gocoap.InternalServerError
+			}
+			return res
+		}
+
+		if m.IsConfirmable() {
+			res.Code = gocoap.Changed
+		}
+		return res
 	}
+}
+
+// Observe method deals with CoAP observe messages.
+func (ca *AdapterService) Observe(auth AuthProvider) MsgHandler {
+	return func(l *net.UDPConn, a *net.UDPAddr, m *gocoap.Message) *gocoap.Message {
+		ca.logger.Log("message", fmt.Sprintf("Got message in Observe: path=%q: %#v from %v", m.Path(), m, a))
+		var res *gocoap.Message
 
-	log.Println("Calling obsTransmit()")
-	log.Println(m.Publisher, m.Protocol, m.Channel, m.Payload)
-	ca.obsTransmit(m)
+		if m.IsConfirmable() {
+			res = &gocoap.Message{
+				Type:      gocoap.Acknowledgement,
+				Code:      gocoap.Content,
+				MessageID: m.MessageID,
+				Token:     m.Token,
+				Payload:   []byte{},
+			}
+			res.SetOption(gocoap.ContentFormat, gocoap.AppJSON)
+		}
+
+		cid := mux.Var(m, channel)
+		_, err := authorize(m, res, cid, auth)
+		if err != nil {
+			ca.logger.Log(err)
+			return res
+		}
+
+		if value, ok := m.Option(gocoap.Observe).(uint32); ok && value == 0 {
+			subject := fmt.Sprintf("channel.%s", cid)
+			if _, err := ca.pubSub.Subscribe(subject, obsHandle(l, a, m, 120)); err != nil {
+				ca.logger.Log("error", fmt.Sprintf("Error occured during subscription to NATS %s", err))
+				res.Code = gocoap.InternalServerError
+				return res
+			}
+			res.AddOption(gocoap.Observe, 0)
+		} else {
```

Comment:
Either handle the condition you mentioned in the comment, or drop the entire branch. No "empty branches" should be present.